### DISCUSSION TITLE
LoadDocument() when document property could be two collections

### DISCUSSION THREAD

**Discussion by tehkyle** [ravendb/ravendb#16854] [Q&A]
We have a "Workflow" document type that has a "_ContentId" field that tracks both Pages and Assets (separate collections) which share a ContentItem base class.

We have a 'WorkflowReportIndex' that is a map reduce and does LoadDocument calls to find this content item and map out of it:
`										  let page = (ContentItem)LoadDocument<Page>(workflow._ContentId)
										  let asset = (ContentItem)LoadDocument<Asset>(workflow._ContentId)
										  let contentItem = page ?? asset`

I've tried a couple variations of this including conditionally calling LoadDocument() from the id type, but each time we are getting the db alert:
Loading documents with mismatched collection name in 'WorkflowReportIndex' index

This appears to be a new alert in 5.4 (we recently upgraded from 5.2) and is tripping our SNMP monitoring.
Are there any suggestions for index definition changes to prevent the warning? or db configuration to suppress it since the index seems to be working normally like it did in 5.2 anyways?

Thanks,
Kyle

**Comment by garayx** [comment]
Hi

In your case you should be fine since you do `let contentItem = page ?? asset` and select the item which is not null.

If you want to get rid of this notification you can do the load in single line like this:

```
let contentItem = (ContentItem)LoadDocument<Page>(workflow._ContentId) ?? (ContentItem)LoadDocument<Asset>(workflow._ContentId) 
```

In this case you don't have `page` & `asset` variables so neither of them is null and you should not get the warning

**Reply by tehkyle** [reply]
Hi

I've updated the index with your suggestion and reset, but after a couple minutes it still creates the warning alert:

![image](https://github.com/ravendb/ravendb/assets/1372402/35064df9-5f0e-4996-80db-417c0021c509)

**Reply by tehkyle** [reply]
Is the LoadDocument collection checking lazy checking the document's metadata after all loads causing this warning?
I tried conditionally loading page or asset and repeating that logic into a test field to confirm:
`let contentItem = (!string.IsNullOrEmpty(workflow._ContentId) && workflow._ContentId[0] == 'x') ? (ContentItem)LoadDocument<Page>(workflow._ContentId) : (ContentItem)LoadDocument<Asset>(workflow._ContentId)`
`let TEST = (!string.IsNullOrEmpty(workflow._ContentId) && workflow._ContentId[0] == 'x') ? "PAGE" : "ASSET"`

and getting:
![image](https://github.com/ravendb/ravendb/assets/1372402/9792a529-80a8-4739-a823-79a3b67e2c1b)
and
![image](https://github.com/ravendb/ravendb/assets/1372402/5917df2f-6a5c-43c1-b98c-e9c4f8913e84)

Which would seem to indicate those asset ids are being properly filtered into calling LoadDocument<Asset>, but the mismatched collection seems to claim they're being loaded with <Page>

**Comment by tehkyle** [comment]
Any assistance would be appreciated as this is flooding our SNMP monitoring.

**Reply by Lwiel** [reply]
Thanks for submitting this, we'll check it this week.

**Comment by ayende** [comment]
Are you able to detect what collection this belongs to based on the document id alone?

**Reply by tehkyle** [reply]
Yes, I've tried doing a id.StartsWith("x") ? LoadDocument<Page> : LoadDocument<Asset>
but still get the warning

**Reply by ayende** [reply]
What are the prefixes that you use?

**Reply by tehkyle** [reply]
for legacy reasons pages are prefixed with 'x'
eg: x123, x20, x33
and assets are 'a/'
eg: a/33, a/1233

**Reply by tehkyle** [reply]
Is there a method of acknowledging and suppressing this alert in the DB so it does not flood our SNMP reporting?

**Reply by Lwiel** [reply]
The code you provided in first post generates load mismatch notification because the check is scoped for a single let clause.
I tried recreating the issue with both converting the code to a single let clause, and checking the document id prefix, however none of them generates such notification for me. Here's the code I used to test this:
```csharp
using Raven.Client.Documents.Indexes;
using Raven.TestDriver;
using Xunit;

namespace Discussions16854;

public class Program : RavenTestDriver
{
    [Fact]
    public void TestLoadCollectionMismatchNotification()
    {
        using (var store = GetDocumentStore())
        {
            using (var session = store.OpenSession())
            {
                var asset1 = new Asset { Id = "a/1", Name = "Asset1" };
                var page1 = new Page { Id = "x/1", Name = "Page1" };
                
                session.Store(asset1);
                session.Store(page1);
                
                session.SaveChanges();

                var workflow1 = new Workflow() { _ContentId = asset1.Id };
                var workflow2 = new Workflow() { _ContentId = page1.Id };
                
                session.Store(workflow1);
                session.Store(workflow2);
                
                session.SaveChanges();

                var indexWithPrefixCheck = new IndexWithPrefixCheck();
                var indexWithSingleLet = new IndexWithSingleLet();
                
                indexWithPrefixCheck.Execute(store);
                indexWithSingleLet.Execute(store);
                
                WaitForIndexing(store);

                WaitForUserToContinueTheTest(store);
            }
        }
    }

    private class Asset
    {
        public string Id { get; set; }
        public string Name { get; set; }
    }

    private class Page
    {
        public string Id { get; set; }
        public string Name { get; set; }
    }

    private class Workflow
    {
        public string _ContentId { get; set; }
    }

    private class ContentItem
    {
        public string Id { get; set; }
        public string Name { get; set; }
    }

    private class IndexWithPrefixCheck :
        AbstractIndexCreationTask<Workflow>
    {
        public IndexWithPrefixCheck()
        {
            Map = workflows => from workflow in workflows
                let contentItem = (!string.IsNullOrEmpty(workflow._ContentId) && workflow._ContentId[0] == 'x') ? (ContentItem)(object)LoadDocument<Page>(workflow._ContentId) : (ContentItem)(object)LoadDocument<Asset>(workflow._ContentId)
                select new { contentItem.Name };
        }
    }

    private class IndexWithSingleLet : AbstractIndexCreationTask<Workflow>
    {
        public IndexWithSingleLet()
        {
            Map = workflows => from workflow in workflows
                let contentItem = (ContentItem)(object)LoadDocument<Page>(workflow._ContentId) ?? (ContentItem)(object)LoadDocument<Asset>(workflow._ContentId)
                select new { contentItem.Name };
        }
    }
}
```
Let me know if I'm missing something, or I'm doing something incorrectly.

**Reply by tehkyle** [reply]
I currently am seeing the alert with:
let contentItem = (ContentItem)LoadDocument<Page>(workflow._ContentId) ?? (ContentItem)LoadDocument<Asset>(workflow._ContentId)
and was also getting it with ..._ContentId.StartsWith("x"), but will try the string array indexing.

I', testing by compiling + pushing new definitions, dismissing the existing alert, and resetting the index. It can take a minute after the index rebuilds but the alert does reappear.

Our "Page" and "Asset" both inherit from "ContentItem" as well, if that makes a difference.

**Reply by tehkyle** [reply]
Is there maybe a stale alert bug? I tried your test index and don't get the alert, and went as far as duplicating the alerting index with simply a new name, and still don't get an alert on the duplicate, but continue to get the alert on the original index even after deleting and recreating it:
![image](https://github.com/ravendb/ravendb/assets/1372402/cda7dc5b-bb8c-4128-80fa-27e5fc040065)


Additionally after dismissing the alert and disabling the index, it still reappears some time later:
![image](https://github.com/ravendb/ravendb/assets/1372402/e8f8e357-217a-4f55-8968-37ca7c3e942f)

**Reply by garayx** [reply]
can you try to restart and reset both of the indexes ?

**Reply by tehkyle** [reply]
I added "MismatchedReferenceLoad" to "Notifications.FilterOut" and reset, which seems to suppress the alert properly.
On your suggestion, I removed the filter and restarted the DB with the single let line implementation and the alert also no longer resurfaces.
Is there some lingering caching of this alert that would have made validating definition changes difficult?

**Reply by garayx** [reply]
Hi

yes it seems that there is still chance that you might get stale notifications, since they are only sent when you open the studio, in any case single let line implementation should not create an alert.
